[css-writing-modes] Refactor TextFlow to be a struct instead of an enum

nt1m opened the pull request:
#### a760ea6009c1bed9b6bba2e890f6fa1a1fcf3d60
<pre>
[css-writing-modes] Refactor TextFlow to be a struct instead of an enum
<a href="https://bugs.webkit.org/show_bug.cgi?id=275966">https://bugs.webkit.org/show_bug.cgi?id=275966</a>
<a href="https://rdar.apple.com/130787071">rdar://130787071</a>

Reviewed by Alan Baradlay.

Names like InlineEastBlockSouth / InlineWestBlockSouth / etc. are quite confusing.
Refactor TextFlow to be a struct instead.

* Source/WebCore/css/CSSCounterStyle.cpp:
(WebCore::CSSCounterStyle::counterForSystemDisclosureClosed):
(WebCore::CSSCounterStyle::counterForSystemDisclosureOpen):
* Source/WebCore/css/CSSCounterStyle.h:
* Source/WebCore/platform/text/WritingMode.h:
(WebCore::TextFlow::isReversed):
(WebCore::TextFlow::isFlipped):
(WebCore::TextFlow::isVertical):
(WebCore::TextFlow::isFlippedLines):
(WebCore::makeTextFlow):
(WebCore::isVerticalWritingMode):
(WebCore::isFlippedWritingMode):
(WebCore::isFlippedLinesWritingMode):
(WebCore::mapLogicalSideToPhysicalSide):
(WebCore::mapPhysicalSideToLogicalSide):
(WebCore::mapLogicalCornerToPhysicalCorner):
(WebCore::mapPhysicalCornerToLogicalCorner):
(WebCore::mapLogicalAxisToPhysicalAxis):
(WebCore::mapPhysicalAxisToLogicalAxis):
(WebCore::operator&lt;&lt;):
(WebCore::makeTextFlowInitalizer): Deleted.
(): Deleted.
(WebCore::isReversedTextFlow): Deleted.
(WebCore::isFlippedTextFlow): Deleted.
(WebCore::isVerticalTextFlow): Deleted.
(WebCore::isFlippedLinesTextFlow): Deleted.
* Source/WebCore/rendering/RenderCounter.cpp:
(WebCore::RenderCounter::originalText const):
* Tools/TestWebKitAPI/Tests/WebCore/WritingModeTests.cpp:

Canonical link: <a href="https://commits.webkit.org/280494@main">https://commits.webkit.org/280494@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2301b49771a6eda64f5cff09cfeff4a59a7fbb07

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56801 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36129 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9275 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60417 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7244 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58929 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43752 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7434 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46005 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5071 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58831 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33943 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49020 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26862 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30723 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6351 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6248 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52653 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6622 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62099 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/714 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6727 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53266 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/717 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49075 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53288 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12559 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/605 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31958 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33043 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34128 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32789 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/platform/text/WritingMode.h:
```diff
@@ -75,54 +75,45 @@ constexpr inline BlockFlowDirection writingModeToBlockFlowDirection(WritingMode
-constexpr unsigned TextFlowVerticalMask = 1 << 2;
-
-constexpr inline bool isReversedTextFlow(TextFlow textflow)
+constexpr inline bool isReversedTextFlow(TextFlow flow)
```
reversedTextFlow is just a bad (and LTR focused) name. we should rename it in a follow-up PR.

nt1m replied:
https://bugs.webkit.org/show_bug.cgi?id=276024

alanbaradlay replied:
Thank you!

alanbaradlay commented on Source/WebCore/platform/text/WritingMode.h:
```diff
@@ -75,54 +75,45 @@ constexpr inline BlockFlowDirection writingModeToBlockFlowDirection(WritingMode
 }
 
-constexpr inline bool isFlippedTextFlow(TextFlow textflow)
+constexpr inline bool isFlippedTextFlow(TextFlow flow)
```
this should get a new name too.

fantasai commented on Source/WebCore/platform/text/WritingMode.h:
```diff
@@ -75,54 +75,44 @@ constexpr inline BlockFlowDirection writingModeToBlockFlowDirection(WritingMode
-    InlineNorthBlockWest = makeTextFlowInitalizer(BlockFlowDirection::RightToLeft, TextDirection::RTL)
+struct TextFlow {
+    BlockFlowDirection blockFlowDirection;
+    TextDirection inlineBaseDirection;
```
You need to incorporate `text-orientation` as well. See https://searchfox.org/mozilla-central/source/layout/generic/WritingModes.h#133

nt1m replied:
I think that would make sense if we started using it, but that's not the case atm, so I'll refrain from doing that.

fantasai replied:
I guess we need to fix https://bugs.webkit.org/show_bug.cgi?id=164507 first anyway.

fantasai commented on Source/WebCore/platform/text/WritingMode.h:
```diff
@@ -75,54 +75,44 @@ constexpr inline BlockFlowDirection writingModeToBlockFlowDirection(WritingMode
+
+    // FIXME: Is this correct?
     if (writingMode == WritingMode::SidewaysLr)
         inlineBaseDirection = direction == TextDirection::RTL ? TextDirection::LTR : TextDirection::RTL;
```
This should be a constructor on the TextFlow class.

nt1m replied:
I filed a bug for that https://bugs.webkit.org/show_bug.cgi?id=276028

fantasai commented on Source/WebCore/platform/text/WritingMode.h:
```diff
@@ -75,54 +75,44 @@ constexpr inline BlockFlowDirection writingModeToBlockFlowDirection(WritingMode
+}
 
-constexpr inline bool isReversedTextFlow(TextFlow textflow)
+constexpr inline bool isReversedTextFlow(TextFlow flow)
```
This and all the other helper functions should be methods on the TextFlow class.

fantasai commented on Source/WebCore/platform/text/WritingMode.h:
```diff
@@ -75,66 +75,56 @@ constexpr inline BlockFlowDirection writingModeToBlockFlowDirection(WritingMode
+    {
+        return blockFlowDirection == BlockFlowDirection::LeftToRight
+            || blockFlowDirection == BlockFlowDirection::RightToLeft;
+    }
```
If you set up the BlockFlowDirection constants correctly, you should be able to use a single bit check here, too. That's more efficient than two conditional checks.

nt1m replied:
I'd privilege explicitness over shortness here, I don't think a bit mask is necessarily faster.

fantasai replied:
Bit masking is faster than conditional checks with potentially failed branch prediction.

nt1m replied:
I guess you could cast to an int then bit mask , but I really prefer explicit checks here

fantasai commented on Source/WebCore/platform/text/WritingMode.h:
```diff
@@ -75,66 +75,56 @@ constexpr inline BlockFlowDirection writingModeToBlockFlowDirection(WritingMode
-    InlineNorthBlockWest = makeTextFlowInitalizer(BlockFlowDirection::RightToLeft, TextDirection::RTL)
-};
+struct TextFlow {
+    BlockFlowDirection blockFlowDirection;
```
Since this is a TextFlow object, maybe just `blockDirection` is enough?

fantasai replied:
(Shorter words here make things less noisy in the complex conditionals these often end up in, and also shorter helps avoid awkward wrapping in such conditionals.)

fantasai commented on Source/WebCore/platform/text/WritingMode.h:
```diff
@@ -75,66 +75,56 @@ constexpr inline BlockFlowDirection writingModeToBlockFlowDirection(WritingMode
-};
+struct TextFlow {
+    BlockFlowDirection blockFlowDirection;
+    TextDirection inlineBaseDirection;
```
I think let's stick with textDirection here; it's not a physical inline direction, but more about bidi interpretation...

fantasai reviewed (approved):
I'll leave the question about bit masking vs `||` to Alan. The rest looks good.

nt1m committed:
[css-writing-modes] Refactor TextFlow to be a struct instead of an enum
https://bugs.webkit.org/show_bug.cgi?id=275966
rdar://130787071

Reviewed by Alan Baradlay.

Names like InlineEastBlockSouth / InlineWestBlockSouth / etc. are quite confusing.
Refactor TextFlow to be a struct instead.

* Source/WebCore/css/CSSCounterStyle.cpp:
(WebCore::CSSCounterStyle::counterForSystemDisclosureClosed):
(WebCore::CSSCounterStyle::counterForSystemDisclosureOpen):
* Source/WebCore/css/CSSCounterStyle.h:
* Source/WebCore/platform/text/WritingMode.h:
(WebCore::TextFlow::isReversed):
(WebCore::TextFlow::isFlipped):
(WebCore::TextFlow::isVertical):
(WebCore::TextFlow::isFlippedLines):
(WebCore::makeTextFlow):
(WebCore::isVerticalWritingMode):
(WebCore::isFlippedWritingMode):
(WebCore::isFlippedLinesWritingMode):
(WebCore::mapLogicalSideToPhysicalSide):
(WebCore::mapPhysicalSideToLogicalSide):
(WebCore::mapLogicalCornerToPhysicalCorner):
(WebCore::mapPhysicalCornerToLogicalCorner):
(WebCore::mapLogicalAxisToPhysicalAxis):
(WebCore::mapPhysicalAxisToLogicalAxis):
(WebCore::operator<<):
(WebCore::makeTextFlowInitalizer): Deleted.
(): Deleted.
(WebCore::isReversedTextFlow): Deleted.
(WebCore::isFlippedTextFlow): Deleted.
(WebCore::isVerticalTextFlow): Deleted.
(WebCore::isFlippedLinesTextFlow): Deleted.
* Source/WebCore/rendering/RenderCounter.cpp:
(WebCore::RenderCounter::originalText const):
* Tools/TestWebKitAPI/Tests/WebCore/WritingModeTests.cpp:

Canonical link: https://commits.webkit.org/280494@main
